onnxruntime/core/providers/coreml/builders/impl/gather_op_builder.cc

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/coreml/builders/impl/base_op_builder.h"
+
+#include "core/providers/coreml/builders/op_builder_factory.h"
+#include "core/providers/coreml/shape_utils.h"
+#include "core/providers/shared/utils/utils.h"
+
+#if defined(__APPLE__)
+#include "core/providers/coreml/builders/model_builder.h"
+#endif
+
+namespace onnxruntime::coreml {
+
+class GatherOpBuilder : public BaseOpBuilder {
+  // Add operator related
+#ifdef __APPLE__
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override;
+#endif
+
+  // Operator support related
+ private:
+  bool HasSupportedInputsImpl(const Node& node, const logging::Logger& logger) const override;
+  bool IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                         const logging::Logger& logger) const override;
+};
+
+// Add operator related
+#if defined(__APPLE__)
+namespace {
+int64_t GetAxisAttribute(const Node& node) {
+  NodeAttrHelper node_attr_helper{node};
+  return node_attr_helper.Get("axis", int64_t{0});
+}
+}  // namespace
+
+Status GatherOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                              const logging::Logger& logger) const {
+  auto layer = CreateNNLayer(model_builder, node);
+  layer->mutable_gather()->set_axis(GetAxisAttribute(node));
+  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();    // data
+  *layer->mutable_input()->Add() = node.InputDefs()[1]->Name();    // indices
+  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();  // output
+  model_builder.AddLayer(std::move(layer));
+  return Status::OK();
+}
+#endif  // defined(__APPLE__)
+
+// Operator support related
+bool GatherOpBuilder::HasSupportedInputsImpl(const Node& node, const logging::Logger& logger) const {
+  int32_t input_type;
+  if (!GetType(*node.InputDefs()[0], input_type, logger))
+    return false;
+
+  if (input_type != ONNX_NAMESPACE::TensorProto_DataType_FLOAT &&
+      input_type != ONNX_NAMESPACE::TensorProto_DataType_INT64) {
+    LOGS(logger, VERBOSE) << "[" << node.OpType()
+                          << "] Input type: [" << input_type
+                          << "] is not supported for now";
```

Comment:
Just curious:
Gather Op does not have any computations, Do we need to consider the input dtypes? Can we convert any dtype into int64 and end up converting back?

Reply:
we need to consider types to the extent that CoreML only supports a subset of them. I think we could allow all of the types that it supports, but for now just starting with int64 as that is the output type of ONNX Shape. the default in BaseOpBuilder is just float, so adding int64 in addition to that here.